### PR TITLE
[KYUUBI #6250] Drop support for Spark 3.1

### LINKED ISSUE #6250: [TASK][EASY] Drop support for Spark 3.1
### What's the level of this task?

EASY

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before creating

- [X] I have searched in the [task list](https://github.com/orgs/apache/projects/343) and found no similar tasks.

### Mentor

- [X] I have sufficient expertise on this task, and I volunteer to be a mentor of this task to guide contributors through the task.

### Skill requirements

- Basic knowledge of Scala, Spark, GitHub Actions

### Background and Goals

We have already deprecated the support for Spark 3.1 in 1.9.0(https://github.com/apache/kyuubi/pull/6091), now, we'd like to drop the support for Spark 3.1 and obliterate the related code.

### Implementation steps

- Update code of `kyuubi-spark-sql-engine`
- Update CI(GHA workflow)
- Update docs (quick start, migration guide, etc.)

### Additional context

Introduction of [2024H1 Kyuubi Code Contribution Program](https://github.com/apache/kyuubi/issues/6232)

### PR DESCRIPTION
**Pull request by liuxiaocs7**
# :mag: Description
## Issue References 🔗

This pull request fixes #6250

## Describe Your Solution 🔧

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Test Plan 🧪

Pass CI

---

# Checklist 📝

- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

**Comment by pan3793** [comment]
thanks for making this PR, I noticed there are some places missed, you can search codebase of `kyuubi-spark-sql-engine` module with keywords "3.2"

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala [hunk 1]
```diff
@@ -197,8 +190,7 @@ object SparkDatasetHelper extends Logging {
 
   private def doCommandResultExec(command: SparkPlan): Array[Array[Byte]] = {
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:191]: this method could be further changed to 
```suggestion
  private def doCommandResultExec(commandResult: CommandResultExec): Array[Array[Byte]] = {
```

Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:191]: one more place: SPARK-35798 also applicable for `doLocalTableScan`

Comment (liuxiaocs7) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala]: `doCommandResultExec` and `doLocalTableScan` are similar now, do the common parts need to be extracted?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/FetchOrcStatement.scala [hunk 2]
```diff
@@ -76,27 +75,13 @@ class FetchOrcStatement(spark: SparkSession) {
...
-            classOf[StructType],
-            classOf[Array[Int]])
-          .build[OrcDeserializer]()
-          .newInstance(
-            new StructType,
-            orcSchema,
-            colId)
-      }
+      // SPARK-34535 changed the constructor signature of OrcDeserializer
+      DynConstructors.builder()
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/FetchOrcStatement.scala:79]: we don't need reflection then

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala [hunk 3]
```diff
@@ -223,30 +210,6 @@ object SparkDatasetHelper extends Logging {
...
-      adaptiveSparkPlanExec: AdaptiveSparkPlanExec,
-      fun: SparkPlan => T): T = {
-    val plan = invokeAs[SparkPlan](adaptiveSparkPlanExec, "getFinalPhysicalPlan")
-    val result = fun(plan)
-    invokeAs[Unit](adaptiveSparkPlanExec, "finalPlanUpdate")
-    result
-  }
-
   /**
    * offset support was add since Spark-3.4(set SPARK-28330), to ensure backward compatibility with
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:214]:
```suggestion
   * offset support was add in SPARK-28330(3.4.0), to ensure backward compatibility with
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/FetchOrcStatement.scala (hunk 2 above)
Comment (liuxiaocs7) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/FetchOrcStatement.scala]: done~

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala (hunk 1 above)
Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala]: the current state is fine

**Comment by pan3793** [comment]
LGTM if CI pass

**Review by cfmcgrady — APPROVED** [review]
LGTM

**Pull request closed without merging**